packages/core/src/services/traces/bulkCreateTracesAndSpans.ts

```diff
@@ -215,60 +180,139 @@ async function createTraces(
+                (COALESCE(${spans.links}, '[]'::jsonb) || COALESCE(EXCLUDED.links, '[]'::jsonb))
+              ) AS element
+            )
+          END
+        `,
+        // Other fields remain the same
+        name: sql`COALESCE(${spans.name}, EXCLUDED.name)`,
+        kind: sql`COALESCE(${spans.kind}, EXCLUDED.kind)`,
+        startTime: sql`COALESCE(${spans.startTime}, EXCLUDED.start_time)`,
+        endTime: sql`COALESCE(${spans.endTime}, EXCLUDED.end_time)`,
+        status: sql`COALESCE(${spans.status}, EXCLUDED.status)`,
+        statusMessage: sql`COALESCE(${spans.statusMessage}, EXCLUDED.status_message)`,
+        internalType: sql`COALESCE(${spans.internalType}, EXCLUDED.internal_type)`,
+        model: sql`COALESCE(${spans.model}, EXCLUDED.model)`,
+        input: sql`COALESCE(${spans.input}, EXCLUDED.input)`,
+        output: sql`COALESCE(${spans.output}, EXCLUDED.output)`,
+        inputTokens: sql`
+          CASE
+            WHEN ${spans.inputTokens} = 0 AND EXCLUDED.input_tokens IS NOT NULL AND EXCLUDED.input_tokens != 0
+            THEN EXCLUDED.input_tokens
+            ELSE ${spans.inputTokens}
+          END
+        `,
+        outputTokens: sql`
+          CASE
+            WHEN ${spans.outputTokens} = 0 AND EXCLUDED.output_tokens IS NOT NULL AND EXCLUDED.output_tokens != 0
+            THEN EXCLUDED.output_tokens
+            ELSE ${spans.outputTokens}
+          END
+        `,
+        totalTokens: sql`
+          CASE
+            WHEN ${spans.totalTokens} = 0 AND EXCLUDED.total_tokens IS NOT NULL AND EXCLUDED.total_tokens != 0
+            THEN EXCLUDED.total_tokens
+            ELSE ${spans.totalTokens}
+          END
+        `,
+        inputCostInMillicents: sql`
+          CASE
+            WHEN ${spans.inputCostInMillicents} = 0 AND EXCLUDED.input_cost_in_millicents IS NOT NULL AND EXCLUDED.input_cost_in_millicents != 0
+            THEN EXCLUDED.input_cost_in_millicents
+            ELSE ${spans.inputCostInMillicents}
+          END
+        `,
+        outputCostInMillicents: sql`
+          CASE
+            WHEN ${spans.outputCostInMillicents} = 0 AND EXCLUDED.output_cost_in_millicents IS NOT NULL AND EXCLUDED.output_cost_in_millicents != 0
+            THEN EXCLUDED.output_cost_in_millicents
+            ELSE ${spans.outputCostInMillicents}
+          END
+        `,
+        totalCostInMillicents: sql`
+          CASE
+            WHEN ${spans.totalCostInMillicents} = 0 AND EXCLUDED.total_cost_in_millicents IS NOT NULL AND EXCLUDED.total_cost_in_millicents != 0
+            THEN EXCLUDED.total_cost_in_millicents
+            ELSE ${spans.totalCostInMillicents}
+          END
+        `,
+        documentUuid: sql`COALESCE(${spans.documentUuid}, EXCLUDED.document_uuid)`,
+        distinctId: sql`COALESCE(${spans.distinctId}, EXCLUDED.distinct_id)`,
+        commitUuid: sql`COALESCE(${spans.commitUuid}, EXCLUDED.commit_uuid)`,
+        updatedAt: new Date(),
+      },
```

Comment:
the meat is here, basically merging duplicated spans via "on conflict" postgres handler and a bunch of sql rules

Reply:
A lot of meat here 🤯